client/db/bolt/db.go

```diff
@@ -0,0 +1,535 @@
+			put(quoteKey, uint32Bytes(md.Quote)).
+			put(statusKey, []byte{byte(md.Status)}).
+			put(dexKey, []byte(md.DEX)).
+			put(updateTimeKey, uint64Bytes(timeNow())).
+			put(proofKey, md.Proof.Encode()).
+			put(matchKey, order.EncodeMatch(match)).
+			err()
+	})
+}
+
+// ActiveMatches retrieves the matches that are in an active state, which is
+// any state except order.MatchComplete.
+func (db *boltDB) ActiveMatches() ([]*dexdb.MetaMatch, error) {
+	return db.filteredMatches(func(mBkt *bolt.Bucket) bool {
+		status := mBkt.Get(statusKey)
+		return len(status) != 1 || status[0] != uint8(order.MatchComplete)
+	})
+}
+
+// filteredMatches gets all matches that pass the provided filter function. Each
+// match's bucket is provided to the filter, and a boolean true return value
+// indicates the match should be decoded and returned.
+func (db *boltDB) filteredMatches(filter func(*bolt.Bucket) bool) ([]*dexdb.MetaMatch, error) {
+	var matches []*dexdb.MetaMatch
+	var err error
+	db.matchesView(func(master *bolt.Bucket) error {
+		master.ForEach(func(k, _ []byte) error {
+			mBkt := master.Bucket(k)
+			if mBkt == nil {
+				return fmt.Errorf("match %x bucket is not a bucket", k)
+			}
+			if filter(mBkt) {
+				var match *order.UserMatch
+				var proof *dexdb.MatchProof
+				matchB := mBkt.Get(matchKey)
+				if matchB == nil {
+					return fmt.Errorf("nil match bytes for %x", k)
+				}
+				match, err = order.DecodeMatch(bCopy(matchB))
+				if err != nil {
+					return fmt.Errorf("error decoding match %x: %v", k, err)
+				}
+				proofB := mBkt.Get(proofKey)
+				if len(proofB) == 0 {
+					return fmt.Errorf("empty proof")
+				}
+				proof, err = dexdb.DecodeMatchProof(bCopy(proofB))
+				if err != nil {
+					return fmt.Errorf("error decoding proof: %v", err)
+				}
+				statusB := mBkt.Get(statusKey)
+				if len(statusB) != 1 {
+					return fmt.Errorf("expected status length 1, got %d", len(statusB))
+				}
+				matches = append(matches, &dexdb.MetaMatch{
+					MetaData: &dexdb.MatchMetaData{
+						Proof:  *proof,
+						Status: order.MatchStatus(statusB[0]),
+						DEX:    string(mBkt.Get(dexKey)),
+						Base:   intCoder.Uint32(mBkt.Get(baseKey)),
+						Quote:  intCoder.Uint32(mBkt.Get(quoteKey)),
+					},
+					Match: match,
+				})
+			}
+			return nil
+		})
+		return nil
+	})
+	return matches, err
+}
+
+// matchesView is a convenience function for reading from the match bucket.
+func (db *boltDB) matchesView(f bucketFunc) error {
+	return db.withBucket(matchesBucket, db.View, f)
+}
+
+// matchesUpdate is a convenience function for updating the match bucket.
+func (db *boltDB) matchesUpdate(f bucketFunc) error {
+	return db.withBucket(matchesBucket, db.Update, f)
+}
+
+// makeTopLevelBuckets creates a top-level bucket for each of the provided keys,
+// if the bucket doesn't already exist.
+func (db *boltDB) makeTopLevelBuckets(buckets [][]byte) error {
+	var err error
+	db.Update(func(tx *bolt.Tx) error {
```

Comment:
`bbolt.(*DB).Update` seems to have return paths that return non-nil errors other than the execution of the provided function.